rust/worker/src/utils/convert.rs

```diff
@@ -0,0 +1,165 @@
+            where_clause,
+        })
+    }
+}
+
+impl From<chroma_proto::LimitOperator> for LimitOperator {
+    fn from(value: chroma_proto::LimitOperator) -> Self {
+        Self {
+            skip: value.skip,
+            fetch: value.fetch,
+        }
+    }
+}
+
+impl From<chroma_proto::ProjectionOperator> for ProjectionOperator {
+    fn from(value: chroma_proto::ProjectionOperator) -> Self {
+        Self {
+            document: value.document,
+            embedding: value.embedding,
+            metadata: value.metadata,
+        }
+    }
+}
+
+impl TryFrom<chroma_proto::KnnProjectionOperator> for KnnProjectionOperator {
+    type Error = Status;
+
+    fn try_from(value: chroma_proto::KnnProjectionOperator) -> Result<Self, Status> {
+        Ok(Self {
+            projection: value
+                .projection
+                .ok_or(Status::invalid_argument("Invalid Projection Operator"))?
+                .into(),
+            distance: value.distance,
+        })
+    }
+}
+
+impl TryFrom<ProjectionRecord> for chroma_proto::ProjectionRecord {
+    type Error = Status;
+
+    fn try_from(value: ProjectionRecord) -> Result<Self, Status> {
+        Ok(Self {
+            id: value.id,
+            document: value.document,
+            embedding: value
+                .embedding
+                .map(|embedding| {
+                    let embedding_dimension = embedding.len();
+                    chroma_proto::Vector::try_from((
+                        embedding,
+                        ScalarEncoding::FLOAT32,
+                        embedding_dimension,
+                    ))
+                })
+                .transpose()
+                .map_err(|err| Status::internal(err.to_string()))?,
+            metadata: value.metadata.map(|metadata| metadata.into()),
+        })
+    }
+}
+
+impl TryFrom<ProjectionOutput> for GetResult {
+    type Error = Status;
+
+    fn try_from(value: ProjectionOutput) -> Result<Self, Status> {
+        Ok(Self {
+            records: value
+                .records
+                .into_iter()
+                .map(TryInto::try_into)
+                .collect::<Result<_, _>>()?,
+        })
+    }
+}
+
+impl TryFrom<KnnProjectionRecord> for chroma_proto::KnnProjectionRecord {
+    type Error = Status;
+
+    fn try_from(value: KnnProjectionRecord) -> Result<Self, Status> {
+        Ok(chroma_proto::KnnProjectionRecord {
+            record: Some(value.record.try_into()?),
+            distance: value.distance,
+        })
+    }
+}
+
+impl TryFrom<KnnProjectionOutput> for KnnResult {
+    type Error = Status;
+
+    fn try_from(value: KnnProjectionOutput) -> Result<Self, Status> {
+        Ok(KnnResult {
+            records: value
+                .records
+                .into_iter()
+                .map(TryInto::try_into)
+                .collect::<Result<_, _>>()?,
+        })
+    }
+}
+
+pub fn from_proto_knn(knn: chroma_proto::KnnOperator) -> Result<Vec<KnnOperator>, Status> {
```

Comment:
why not make this TryFrom?

Reply:
We cannot. Neither `chroma_proto::KnnOperator` nor `Vec<KnnOperator>` is owned by our crate so we cannot implement the trait, unless we make our own trait.